src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceGenerator.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// We only build the Source Generator in the netstandard target
+#if NETSTANDARD
+
+#nullable enable
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Xml;
+using System.Xml.Serialization;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace CSharpSyntaxGenerator
+{
+    [Generator]
+    public sealed class SourceGenerator : ISourceGenerator
+    {
+        public void Initialize(InitializationContext context)
+        {
+        }
+
+        public void Execute(SourceGeneratorContext context)
+        {
+            var syntaxXml = context.AdditionalFiles.SingleOrDefault(a => Path.GetFileName(a.Path) == "Syntax.xml");
+
+            if (syntaxXml == null)
+            {
+                return;
+            }
+
+            var syntaxXmlText = syntaxXml.GetText();
+
+            if (syntaxXmlText == null)
+            {
+                // TODO: report a diagnostic here
+                return;
+            }
+
+            var reader = XmlReader.Create(new SourceTextReader(syntaxXmlText), new XmlReaderSettings { DtdProcessing = DtdProcessing.Prohibit });
+            var serializer = new XmlSerializer(typeof(Tree));
+            var tree = (Tree)serializer.Deserialize(reader);
+
+            TreeFlattening.FlattenChildren(tree);
+
+            AddResult(writer => SourceWriter.WriteMain(writer, tree, context.CancellationToken), "Syntax.xml.Main.Generated.cs");
+            AddResult(writer => SourceWriter.WriteInternal(writer, tree, context.CancellationToken), "Syntax.xml.Internal.Generated.cs");
+            AddResult(writer => SourceWriter.WriteSyntax(writer, tree, context.CancellationToken), "Syntax.xml.Syntax.Generated.cs");
+
+            void AddResult(Action<TextWriter> writeFunction, string hintName)
+            {
+                // Write out the contents to a StringBuilder to avoid creating a single large string
+                // in memory
+                var stringBuilder = new StringBuilder();
+                using (var textWriter = new StringWriter(stringBuilder))
+                {
+                    writeFunction(textWriter);
+                }
+
+                // And create a SourceText from the StringBuilder, once again avoiding allocating a single massive string
+                context.AddSource(hintName, SourceText.From(new StringBuilderReader(stringBuilder), stringBuilder.Length, encoding: Encoding.UTF8));
```

Comment:
@chsienki Should we have an API that lets you do AddSource and directly give it a StringBuilder? Having to write this myself seems icky.

Reply:
Why not use the overload that takes a string?

```suggestion
                context.AddSource(hintName, stringBuilder.ToString());
```

Reply:
@sharwell: because that'll allocate a 2MB string. :smile:

Reply:
@jasonmalinowski I'm 100% fine with adding at string builder overload. Do you want me to add it to this PR?

Reply:
I'd say no, so if we have to revert this we're also not reverting a public API change that somebody else might depend on.

Reply:
Filed https://github.com/dotnet/roslyn/issues/49082 to track it.

Reply:
@jasonmalinowski 
In #49689, the conclusion was that there shouldn't be a StringBuilder overload because it's mutable, but it seems like being able to create a SourceText from a TextReader like this has the exactly the same problem?

Reply:
Actually no it doesn't, because `SourceText.From(TextReader)` calls `ReadToEnd` and allocates the string anyway, so this doesn't really avoid the allocation 😄
https://sourceroslyn.io/#Microsoft.CodeAnalysis/Text/SourceText.cs,121